gsmphys/mfpbltq.f

```diff
@@ -99,14 +105,35 @@ subroutine mfpbltq(im,ix,km,kmpbl,ntcw,ntrac1,delt,
         endif
       enddo
 !
+! kgao 12/08/2023: adjust entrainment/detrainment rate based on pbl-mean tke  
+!
+!  if tkemean>tkcrt, ce0t=sqrt(tkemean/tkcrt)*ce0
+!
+      do i=1,im
+        if( use_tke_ent_det .and. cnvflg(i) ) then
+          ce0t(i) = ce0 * vez0fun(i)
+          if(tkemean(i) > tkcrt) then
+            tem = sqrt(tkemean(i)/tkcrt)
+            tem1 = min(tem, cmxfac)
+            tem2 = tem1 * ce0
+            ce0t(i) = max(ce0t(i), tem2)
+          endif
+        endif
+      enddo
+!
 !  compute entrainment rate
 !
       do i=1,im
         if(cnvflg(i)) then
           k = kpbl(i) / 2
           k = max(k, 1)
           delz(i) = zl(i,k+1) - zl(i,k)
-          xlamax(i) = ce0 / delz(i)
+          ! kgao 12/08/2023
+          if (use_tke_ent_det) then
+             xlamax(i) = ce0t(i) / delz(i)
+          else
+             xlamax(i) = ce0 / delz(i)
+          endif 
```

Comment:
Since the only difference is whether to use ce0t(i) or ce0 based on `use_tke_ent_det==T/F` you could have set ce0t(i)==ce0 above for the case where it is `False` and not had to use the if-test here and in similar loops below.